File: c/12.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#include <math.h>
#define defaultInput "../../Inputs/2021_12.txt"
#include "../../Modules/input.h"
#include "../../Modules/dict.h"


int maxVal(struct dict *d) {
    struct vector *vals = values2vector(d);
    int max = *(int*)vals->arr[0];
    for (int i = 1; i < vals->len; i++) {
        if (*(int*)vals->arr[i] > max) {
            max = *(int*)vals->arr[i];
        }
    }

    return max;
}


int findPaths(struct dict *neighborData, struct dict *visitedData, char *node, bool p2) {
    if (strncmp(node, "end", 3) == 0) {
        return 1;
    }

    int *nodeVisitTimes = (int*)getDictVal(visitedData, node);
    if (node[0] == tolower(node[0])) {
        *nodeVisitTimes = *nodeVisitTimes + 1;
    }

    struct vector *neighbors = (struct vector *)getDictVal(neighborData, node);
    int paths = 0;
    for (int i = 0; i < neighbors->len; i++) {
        char *n = (char*)neighbors->arr[i];
        if (strncmp(n, "start", 5) == 0) {
            continue;
        }

        int *neighborVisitTimes = (int*)getDictVal(visitedData, n);
        bool canVisit = (*neighborVisitTimes == 0) || (p2 && maxVal(visitedData) == 1);

        if (canVisit) {
            paths += findPaths(neighborData, visitedData, n, p2);
        }
    }

    if (node[0] == tolower(node[0])) {
        *nodeVisitTimes = *nodeVisitTimes - 1;
    }

    return paths;
}

int part1(char *fileName) {
    struct vector *input_data = multiLine(fileName);
    struct dict *neighborData = createDict(stringsize, sizeofVector, createCopyVector), *visitedData = createDict(stringsize, intsize, copyElement);
    
    for (int i = 0; i < input_data->len; i++) {
        char *line = (char*)calloc(strlen((char*)input_data->arr[i]), sizeof(char));
        strcpy(line, input_data->arr[i]);
        
        char *p1 = strtok(line, "-"), *p2 = strtok(NULL, "-");
        
        if (!inDict(neighborData, p1)) {
            struct vector *nVec = createVector(stringsize, copyElement);
            addDict(neighborData, p1, nVec);
        }

        if (!inDict(neighborData, p2)) {
            struct vector *nVec = createVector(stringsize, copyElement);
            addDict(neighborData, p2, nVec);
        }
        
        struct vector *p1Vec = (struct vector *)getDictVal(neighborData, p1), *p2Vec = (struct vector *)getDictVal(neighborData, p2);
        
        appendVector(p1Vec, p2);
        appendVector(p2Vec, p1);
    }

    struct vector *key_vec = keys2vector(neighborData);
    for (int i = 0; i < key_vec->len; i++) {
        char *key = (char*)key_vec->arr[i];
        int *times = (int*)calloc(1, sizeof(int));
        addDict(visitedData, key, times);
    }

    return findPaths(neighborData, visitedData, "start", false);
}

int part2(char *fileName) {
    struct vector *input_data = multiLine(fileName);
    struct dict *neighborData = createDict(stringsize, sizeofVector, createCopyVector), *visitedData = createDict(stringsize, intsize, copyElement);

    for (int i = 0; i < input_data->len; i++) {
        char *line = (char*)calloc(strlen((char*)input_data->arr[i]), sizeof(char));
        strcpy(line, input_data->arr[i]);

        char *p1 = strtok(line, "-"), *p2 = strtok(NULL, "-");
        
        if (!inDict(neighborData, p1)) {
            struct vector *nVec = createVector(stringsize, copyElement);
            addDict(neighborData, p1, nVec);
        }

        if (!inDict(neighborData, p2)) {
            struct vector *nVec = createVector(stringsize, copyElement);
            addDict(neighborData, p2, nVec);
        }
        
        struct vector *p1Vec = (struct vector *)getDictVal(neighborData, p1), *p2Vec = (struct vector *)getDictVal(neighborData, p2);
        
        appendVector(p1Vec, p2);
        appendVector(p2Vec, p1);
    }

    struct vector *key_vec = keys2vector(neighborData);
    for (int i = 0; i < key_vec->len; i++) {
        char *key = (char*)key_vec->arr[i];
        int *times = (int*)calloc(1, sizeof(int));
        addDict(visitedData, key, times);
    }

    return findPaths(neighborData, visitedData, "start", true);
}


int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

    clock_t t;
    t = clock(); 
    int p1 = part1(inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nNumber of paths visiting small caves at most once: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock(); 
    int p2 = part2(inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nNumber of paths visiting small caves at most twice: %d\nRan in %f seconds\n", p2, t_p2);

    return 0;
}

File: python/12.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
from collections import defaultdict


def part1(data):
    """ 2021 Day 12 Part 1

    >>> part1(['start-A', 'start-b', 'A-c', 'A-b', 'b-d', 'A-end', 'b-end'])
    10
    """

    caves = defaultdict(lambda: set())
    for line in data:
        line = line.split('-')
        
        caves[line[0]].add(line[1])
        caves[line[1]].add(line[0])

    return findPaths('start', caves, defaultdict(lambda: 0), False)


def part2(data):
    """ 2021 Day 12 Part 2

    >>> part2(['start-A', 'start-b', 'A-c', 'A-b', 'b-d', 'A-end', 'b-end'])
    36
    """

    caves = defaultdict(lambda: set())
    for line in data:
        line = line.split('-')
        
        caves[line[0]].add(line[1])
        caves[line[1]].add(line[0])

    return findPaths('start', caves, defaultdict(lambda: 0), True)


def findPaths(curr, caves, visited, p2):
    if curr == "end":
        return 1

    if curr.islower():
        visited[curr] += 1

    paths = 0
    for n in list(caves[curr]):
        if n == "start":
            continue

        if visited[n] == 0 or (p2 and max(visited.values()) == 1):
            paths += findPaths(n, caves, visited, p2)

    if curr.islower():
        visited[curr] -= 1

    return paths


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nNumber of paths visiting small caves once: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nNumber of paths visiting small caves twice: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)